plugin/plugin_test.go

```diff
@@ -0,0 +1,327 @@
+			parser: flow.NewParser,
+		},
+		{
+			desc: "replace input modifier",
+			src:  inputModifierPlugin,
+			parser: func() *flow.Parser {
+				p := flow.NewParser()
+				p.LoadInputModifier("input_modifier", nil)
+				return p
+			},
+			summary: Summary{Replaced: []string{"input_modifier"}},
+		},
+		{
+			desc:             "wrong input format type",
+			src:              "package main\nfunc InputFormats() {}\n",
+			applyErrContains: "failed to assert type",
+			parser:           flow.NewParser,
+		},
+		{
+			desc:   "new input format",
+			src:    inputFormatPlugin,
+			parser: flow.NewParser,
+		},
+		{
+			desc: "replace input format",
+			src:  inputFormatPlugin,
+			parser: func() *flow.Parser {
+				p := flow.NewParser()
+				p.LoadInputFormat("input_format", nil)
+				return p
+			},
+			summary: Summary{Replaced: []string{"input_format"}},
+		},
+		{
+			desc:   "new output format",
+			src:    outputFormatPlugin,
+			parser: flow.NewParser,
+		},
+		{
+			desc: "replace output format",
+			src:  outputFormatPlugin,
+			parser: func() *flow.Parser {
+				p := flow.NewParser()
+				p.LoadOutputFormat("output_format", nil)
+				return p
+			},
+			summary: Summary{Replaced: []string{"output_format"}},
+		},
+		{
+			desc:   "new output modifier",
+			src:    outputModifierPlugin,
+			parser: flow.NewParser,
+		},
+		{
+			desc: "replace output modifier",
+			src:  outputModifierPlugin,
+			parser: func() *flow.Parser {
+				p := flow.NewParser()
+				p.LoadOutputModifier("output_modifier", nil)
+				return p
+			},
+			summary: Summary{Replaced: []string{"output_modifier"}},
+		},
+		{
+			desc:   "new sink",
+			src:    sinkPlugin,
+			parser: flow.NewParser,
+		},
+		{
+			desc: "replace sink",
+			src:  sinkPlugin,
+			parser: func() *flow.Parser {
+				p := flow.NewParser()
+				p.LoadSink("sink", nil)
+				return p
+			},
+			summary: Summary{Replaced: []string{"sink"}},
+		},
+	}
+	for _, c := range cases {
+		c := c
+		t.Run(c.desc, func(t *testing.T) {
+			ctx := session.NewContexts(&session.Config{Debug: true}, nil, nil)
+
+			dir, err := ioutil.TempDir("", "")
+			assert.NilError(t, err)
+			defer os.RemoveAll(dir)
+
+			err = ioutil.WriteFile(path.Join(dir, "main.go"), []byte(c.src), os.ModePerm)
+			assert.NilError(t, err)
+
+			wd, err := os.Getwd()
+			assert.NilError(t, err)
+			goModFile := fmt.Sprintf(goModFileFormat, ksuid.New().String(), path.Join(wd, ".."))
+
+			err = ioutil.WriteFile(path.Join(dir, "go.mod"), []byte(goModFile), os.ModePerm)
+			assert.NilError(t, err)
+
+			cmd := exec.Command("go", "mod", "download", "-x")
+			cmd.Dir = dir
+			out, err := cmd.CombinedOutput()
+			if len(out) > 0 {
+				fmt.Println(string(out))
+			}
+			assert.NilError(t, err)
+
+			pluginFileName := fmt.Sprintf("plugin_%s.so", ksuid.New().String())
+
+			// Build the plugin with the race detector so we can also
+			// run the tests with it as well. Plugins must be built the same
+			// exact way as the tests to be loaded.
```

Comment:
I was wondering how we might test this, but including the source and building the test plugins during the test is clever and pretty straightforward. Nice work!